android/tangram/src/com/mapzen/tangram/MapData.java

```diff
@@ -1,161 +1,154 @@
         return this;
     }
 
     /**
-     * Add a line geometry to this data source
-     * @param line List of LngLat points comprising the line
-     * @return This object, for chaining
+     * Add a polygon feature to this collection.
+     * @param polygon A list of rings describing the shape of the feature. Each
+     * ring is a list of coordinates. The first ring is taken as the "exterior" of the polygon and
+     * rings with opposite winding are considered "holes".
+     * @param properties The properties of the feature, used for filtering and styling according to
+     * the scene file used by the map; may be null.
+     * @return This object, for chaining.
      */
-    public MapData addLine(Properties props, List<LngLat> line) {
-        // need to concatenate points
-        Coordinates coords = new Coordinates();
-        for (LngLat point : line) {
-            coords.add(point);
-        }
-        addLineJNI(props, coords);
-        return this;
-    }
-
-    public MapData addLine(Properties props, Coordinates line) {
-        addLineJNI(props, line);
+    public MapData addPolygon(List<List<LngLat>> polygon, Map<String, String> properties) {
+        features.add(new Polygon(polygon, properties));
         return this;
     }
 
     /**
-     * Add a polygon geometry to this data source
-     * @param polygon List of lines of LngLat points, where each line represents
-     *                a ring in the polygon as described in the GeoJSON spec
-     * @return This object, for chaining
+     * Add features described in a GeoJSON string to this collection.
+     * @param data A string containing a <a href="http://geojson.org/">GeoJSON</a> FeatureCollection
+     * @return This object, for chaining.
      */
-    public MapData addPolygon(Properties props, List<List<LngLat>> polygon) {
-        Polygon poly = new Polygon();
-
-        // for (List<LngLat> ring : polygon) {
-        //     Coordinates out = new Coordinates();
-        //     for (LngLat point : ring) {
-        //         out.add(point);
-        //     }
-        //     poly.add(out);
-        // }
-
-        // TODO add method to add empty ring and get handle to it.
-        Coordinates dummy = new Coordinates();
-        int rings = 0;
-
-        for (List<LngLat> ring : polygon) {
-            poly.add(dummy);
-
-            Coordinates out = poly.get(rings);
-
-            for (LngLat point : ring) {
-                out.add(point);
-            }
-            rings++;
-        }
-        addPolyJNI(props, poly);
+    public MapData addGeoJson(String data) {
+        geojson.add(data);
         return this;
     }
 
-    public MapData addPolygon(Properties props, Polygon polygon) {
-        addPolyJNI(props, polygon);
+    /**
+     * Remove all features from this collection.
+     * @return This object, for chaining.
+     */
+    public MapData clear() {
+        geojson.clear();
+        features.clear();
```

Comment:
Example for a usecase for this be the routing icon on EM:
* get new location
* clear routeIcon MapData
* add new location Point to routeIcon MapData
* syncWithMap